service_configuration_lib/cached_view.py

```diff
@@ -0,0 +1,193 @@
+            watch_manager=watch_manager,
+            default_proc_fun=handler,
+        )
+
+        watch_manager.add_watch(
+            path=self._configs_folder,
+            mask=pyinotify.IN_MOVED_TO | pyinotify.IN_CREATE | pyinotify.IN_DELETE,
+            rec=True,
+            auto_add=True,
+            exclude_filter=self._exclude_filter,
+        )
+        for root, _, files in os.walk(self._configs_folder):
+            for path in files:
+                self._maybe_add_path_to_cache(os.path.join(root, path))
+
+    def _exclude_filter(self, path: str) -> bool:
+        """Directories (not files) to exclude from watching.
+
+        Attempting to watch tmp directories causes pyinotify to log an error,
+        so let's just filter them out.
+        Also it will exclude all folders which don't match with services_names param.
+        """
+        folder_name = Path(path).name
+
+        for exclude_folder in self._exclude_folders_filters:
+            if fnmatch.fnmatch(folder_name, exclude_folder):
+                return True
+
+        for service_name in self._services_names:
+            if fnmatch.fnmatch(folder_name, service_name):
+                return False
+
+        return True
+
+    def _maybe_add_path_to_cache(self, path: str) -> None:
+        result = self._service_name_and_config_from_path(path)
+        if result is None:
+            return
+        service_name, config_name, config_suffix = result
+        log.info(f'{config_name}: Config changed for {service_name}')
+        self.configs_view.add(path, service_name, config_name, config_suffix)
+
+    def _maybe_remove_path_from_cache(self, path: str) -> None:
+        result = self._service_name_and_config_from_path(path)
+        if result is None:
+            return
+        service_name, config_name, config_suffix = result
+        log.info(f'{config_name}: Removing config for {service_name}')
+        self.configs_view.remove(path, service_name, config_name, config_suffix)
+
+    def _service_name_and_config_from_path(self, path: str) -> Optional[Tuple[str, str, str]]:
+        """Convert a config file path to a service name and config name.  For example,
+        `/nail/etc/services/foo/smartstack.yaml` would be converted to `foo` with a
+        config name of `smartstack`
+
+        Returns `None` if path is invalid or we don't care about the file, e.g. not in self._config_names
+
+        :param path: the config file path e.g. `/nail/etc/services/foo/smartstack.yaml`
+        :return: a tuple of the service name, config and suffix e.g. (`foo`, `smartstack`, `yaml`),
+        or `None` if the path is invalid
+        """
+        relpath = os.path.relpath(path, self._configs_folder)
+        filename = Path(relpath)
+
+        # when len == 2, thatss regular case e.g. service/config.yaml
+        # when len == 3, that's case with secrets folder, e.g. service/secrets/config.json
+        if len(filename.parts) == 2 or len(filename.parts) == 3:
```

Comment:
This feels weird to me, and makes me think this shouldn't be in the cacher.

What if the raw filename was just returned and left to the caller to decide what to do with it?

Or, with all this config/service/file stuff, what if the caller was just responsible to provide a regex?

Reply:
Right, that's a bit weird.
The idea was to simplify service and config-names parsing from filename for library users.
It already passing full path, service_name and config_name to user's callbacks.
Perhaps I can have just assumption with len(filename.parts) > 1 and do same service_name and config_name extraction. Or do extraction when I can and put None for service_name into result when the files appeared on the top-level.